### ISSUE TITLE
Latest msgpack (1.4.3) broken on JRuby when using Java 8 VM

### ISSUE THREAD

**Issue by ivoanjo** [msgpack/msgpack-ruby#239]
Hello! I work on Datadog's [ddtrace gem](https://github.com/DataDog/dd-trace-rb) and noticed this issue [on our CI](https://app.circleci.com/pipelines/github/DataDog/dd-trace-rb/5167/workflows/1808cc8b-e055-479c-a50d-f00469b260fb/jobs/193173) today.

The latest msgpack version, when run with the Java 8 VM breaks:

```
$ docker run jruby:9.3-jre8 bash -c "gem install msgpack; ruby -rmsgpack -e 'puts RUBY_DESCRIPTION; [1,2,3].to_msgpack'"
/opt/jruby/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::AF_INET6
/opt/jruby/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::IPPROTO_ICMPV6
/opt/jruby/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::IPPROTO_IPV6
/opt/jruby/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::PF_INET6
/opt/jruby/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::SOL_IPV6
Successfully installed msgpack-1.4.3-java
1 gem installed
jruby 9.3.3.0 (2.6.8) 2022-01-19 b26de1f5c5 OpenJDK 64-Bit Server VM 25.312-b07 on 1.8.0_312-b07 +jit [linux-x86_64]
Unhandled Java exception: java.lang.NoSuchMethodError: java.nio.ByteBuffer.clear()Ljava/nio/ByteBuffer;
java.lang.NoSuchMethodError: java.nio.ByteBuffer.clear()Ljava/nio/ByteBuffer;
           readRubyString at org/msgpack/jruby/Encoder.java:70
                   encode at org/msgpack/jruby/Encoder.java:81
                    write at org/msgpack/jruby/Packer.java:128
                     call at org/msgpack/jruby/Packer$INVOKER$i$1$0$write.gen:-1
             cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:372
                     call at org/jruby/runtime/callsite/CachingCallSite.java:175
              processCall at org/jruby/ir/interpreter/InterpreterEngine.java:316
```

Note that I've used JRuby 9.3 above, but it also happens on 9.2, since the issue is with the version of the Java VM:

```
$ docker run jruby:9.2-jre8 bash -c "gem install msgpack; ruby -rmsgpack -e 'puts RUBY_DESCRIPTION; [1,2,3].to_msgpack'"
Successfully installed msgpack-1.4.3-java
1 gem installed
jruby 9.2.20.1 (2.5.8) 2021-11-30 2a2962fbd1 OpenJDK 64-Bit Server VM 25.312-b07 on 1.8.0_312-b07 +jit [linux-x86_64]
Unhandled Java exception: java.lang.NoSuchMethodError: java.nio.ByteBuffer.clear()Ljava/nio/ByteBuffer;
java.lang.NoSuchMethodError: java.nio.ByteBuffer.clear()Ljava/nio/ByteBuffer;
           readRubyString at org/msgpack/jruby/Encoder.java:70
                   encode at org/msgpack/jruby/Encoder.java:81
                    write at org/msgpack/jruby/Packer.java:128
                     call at org/msgpack/jruby/Packer$INVOKER$i$1$0$write.gen:-1
             cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:376
                     call at org/jruby/runtime/callsite/CachingCallSite.java:175
              processCall at org/jruby/ir/interpreter/InterpreterEngine.java:316
```

This test works correctly if instead of the `jruby:9.2-jre8` and `jruby:9.3-jre8` we switch to `jruby:9.2-jre11` and `jruby:9.3-jre11`.

My analysis of the issue is as follows:

In `Encoder.java` the code calls `buffer.clear()`:

https://github.com/msgpack/msgpack-ruby/blob/f706c95e080fd88209dc4c446a0f2ff0436dd203/ext/java/org/msgpack/jruby/Encoder.java#L70

The `buffer` here refers to a `java.nio.ByteBuffer`. In Java 8, `ByteBuffer::clear` is actually inherited from its parent class:

![Screen Shot 2022-01-20 at 12 25 37](https://user-images.githubusercontent.com/2785847/150338430-299588ef-4d5a-4ac8-95ca-afc3715b93ee.png)

Looking at its parent class, the signature for `clear` is:

![Screen Shot 2022-01-20 at 12 27 21](https://user-images.githubusercontent.com/2785847/150338566-6437a799-5729-44fb-b1c9-c4a9195e0d54.png)

and thus in `ByteBuffer` the signature for calling this method using Java bytecode is

`java.nio.ByteBuffer.clear()Ljava/nio/Buffer;`

...whereas msgpack seems to be looking for a

`java.nio.ByteBuffer.clear()Ljava/nio/ByteBuffer;`

The latter method was actually added in Java 9:

![Screen Shot 2022-01-20 at 12 29 30](https://user-images.githubusercontent.com/2785847/150338917-7349a227-794b-4ab3-939c-ec1c1b4f33de.png)

Notice how it says that it overrides the method `clear`, and the return is now `ByteBuffer`, not just `Buffer`. (Having a subclass override a return with a more specific type is allowed in Java).

Now it is true that msgpack is compiled with source and target version 1.8 (aka Java 8):

https://github.com/msgpack/msgpack-ruby/blob/1af1773974c433ebf786165e18e8991810335d77/Rakefile#L37-L38

But I'm betting that the release was not compiled on Java 8, and that we're running into this known limitation (from [here](https://docs.oracle.com/javase/7/docs/technotes/tools/windows/javac.html#crosscomp-example)):

> If you do not specify the correct version of bootstrap classes, the compiler will use the old language rules (in this example, it will use version 1.6 of the Java programming language) combined with the new bootstrap classes, which can result in class files that do not work on the older platform (in this case, Java SE 6) because reference to non-existent methods can get included.

TL;DR it's not enough to use source and target as Java 1.8, you actually need to be running Java 8 OR feed the compiler with a Java 8 `rt.jar` file.

**Comment by tagomoris** [comment]
Does it mean every laptop for the release operation should run Java 8 at those time? Hmm...
Building `.gem` on CI (w/ Java 8 probably?) is an option. I'm considering other options.

**Comment by ivoanjo** [comment]
> Does it mean every laptop for the release operation should run Java 8 at those time?

Yeah, that would be the most reliable option.

> Building .gem on CI (w/ Java 8 probably?) is an option. I'm considering other options.

Running the build in docker would also work (assuming, of course, you want to and can run Docker on your machine). Or any other Linux VM; I say docker because if you already have it installed it's probably the easiest way.

**Comment by ahorek** [comment]
unfortunately, in rare cases, java 9+ binaries aren't backward compatible with java 8 even in java 8 compatibility mode

building a gem on Java 8 is one option, but there's a workaround https://github.com/msgpack/msgpack-ruby/pull/240

**Comment by tagomoris** [comment]
@ivoanjo I've pushed a pre-release version `1.4.4.pre1` with a compile option for older Java platforms.
Could you try it in your environment?
https://rubygems.org/gems/msgpack/versions/1.4.4.pre1-java

**Comment by ivoanjo** [comment]
Thanks for the amazingly fast turnaround! I can confirm this works fine on JRuby 9.3 and latest 9.2 for dd-trace-rb. I did see an issue with older 9.2 releases, but I'll file a separate ticket for it, since it's a different issue.

**Comment by ivoanjo** [comment]
Actually, scratch that. I found that #228 breaks JRuby 9.2.0.0, but works from 9.2.1.0 on.

So... I don't think it's worth opening an issue for; I'll leave a note on the PR, but if someone out there is running 9.2.0.0 in 2022: please, please don't :)

**Comment by tagomoris** [comment]
This change will be updated and v1.4.4 will be shipped after the release of rake-compiler.
https://github.com/msgpack/msgpack-ruby/pull/241

**Comment by ahorek** [comment]
well, it shouldn't be a big deal for JRuby 9.2.0.0 users to upgrade to any supported version in the 9.2.x.x line. We can add a compatibility switch, but I agree it's not worth it just to support this particular version only. There's really no rational reason to stay on that version these days.

**Comment by tagomoris** [comment]
I've released `v1.4.4` which should fix this issue.

**Comment by ivoanjo** [comment]
Happy with the fix, I think this can be closed now! :)
